syntax = "proto3";

package telegraph;

message Empty {}

message Group {
    int32 id = 1; // optional
    string name = 2;
    string pretty = 3;
    string desc = 4;

    string schema = 5;
    int32 version = 6;

    repeated Node children = 7;
}

message Variable {
    int32 id = 1; // optional
    string name = 3;
    string pretty = 4;
    string desc = 5;

    Type data_type = 6;
}

message Action {
    int32 id = 1; // optional
    string name = 3;
    string pretty = 4;
    string desc = 5;

    Type arg_type = 6;
    Type ret_type = 7;
}

message Node {
    oneof node {
        Group group = 1;
        Variable var = 2;
        Action action = 3;
    }
}

// ------------ Type/Value Utilities -----------------

message Type {
    enum Class {
        INVALID = 0;
        NONE = 1;
        ENUM = 2;
        BOOL = 3;
        UINT8 = 4;
        UINT16 = 5;
        UINT32 = 6;
        UINT64 = 7;
        INT8 = 8;
        INT16 = 9;
        INT32 = 10;
        INT64 = 11;
        FLOAT = 12;
        DOUBLE = 13;
    }
    Class type =  1;
    string name = 2;
    repeated string labels = 3;
}

message Value {
    oneof type {
        Empty none = 1;
        int32 en = 2;
        bool b = 3;
        uint32 u8 = 4;
        uint32 u16 = 5;
        uint32 u32 = 6;
        uint64 u64 = 7;
        int32  i8 = 8;
        int32  i16 = 9;
        int32  i32 = 10;
        int64  i64 = 11;
        float  f = 12;
        double d = 13;
    }
}

message Datapoint {
    uint64 timestamp = 1;
    Value value = 2;
}
